CORE-5090: Add ConsensualStatesLedger implementation + test app

omfR3 opened the pull request:
This PR adds an implementation of the `ConsensualStatesLedger` interface, intended to be injected into flows. Currently it has no meaningful logic and just doubles the input value so we can test that it can be hooked up in a simple test app.

Linked with [this interface PR](https://github.com/corda/corda-api/pull/441) in corda-api.

## TODO
* [X] Basic implementation + OSGi/bnd boilerplate
* [X] Dummy test app
* [x] Move some existing components to a common place so they can be reused in integration tests for flows on a fake virtual node (virtual virtual node?)
* [x] Actually works
* [ ] ~Use `VirtualNodeService` to start a flow~
* [x] Make test CPB work

## Current status

Integration test succeeds but tests nothing other than loading the injected service. Couldn't get the virtual node flow to work because it needs a lot more mocked out infra (e.g. a fake in-memory message bus instead of Kafka).
Triggering an incredibly basic flow with the test CPB running in a local Kubernetes cluster now succeeds.

omfR3 committed:
CORE-5090: Add the beginnings of an injectable `ConsensualStatesLedger` service + simple app for E2E testing

omfR3 commented on .gitignore:
```diff
@@ -130,3 +130,4 @@ node_modules/
 /testing/soak/GroupPolicy.json
 /testing/soak/*.cpi
 /testing/soak/*.csv
+/values.yaml
```
^ Added to avoid accidentally committing this after local testing.

omfR3 committed:
Add ledger packages to sandbox whitelist

omfR3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextComponentImpl.kt:
```diff
@@ -65,6 +65,7 @@ class SandboxGroupContextComponentImpl @Activate constructor(
                 "net.corda.crypto",
                 "net.corda.kotlin-stdlib-jdk7.osgi-bundle",
                 "net.corda.kotlin-stdlib-jdk8.osgi-bundle",
+                "net.corda.ledger",
```
@driessamyn @lankydan @JamesHR3 sorry to bug you about this, but I'm getting nowhere fast with this and thought your spidey senses might help. I added the ledger package to the sandbox whitelist, hoping to get around this error:
```
org.osgi.framework.BundleException: Unable to resolve net.corda.consensual-ledger [135](R 135.0):
  missing requirement [net.corda.consensual-ledger [135](R 135.0)] osgi.wiring.package;
    (&(osgi.wiring.package=net.corda.v5.ledger.models)(version>=5.0.0)(!(version>=6.0.0)))
    Unresolved requirements: [[net.corda.consensual-ledger [135](R 135.0)] osgi.wiring.package;
      (&(osgi.wiring.package=net.corda.v5.ledger.models)(version>=5.0.0)(!(version>=6.0.0)))]
	at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4398) ~[flow-worker.jar:?]
```
...but it didn't help. Actually I don't quite understand what the error message is saying -- is it having trouble finding the `net.corda.consensual-ledger` module (or bundle, or ...?) too, or specifically having a problem with the interface I added in `corda-api` (in the new `net.corda.v5.ledger.models` package)? What am I missing that would cause it to not see the interface in corda-api?

JamesHR3 replied:
The thing that's missing is always the thing before the `Unresolved requirements` text (yes this isn't super helpful). Here it's saying that `net.corda.v5.ledger.models` can't be found.

You might need to tell the application packaging code to add your bundle(s) to the flow worker by declaring a `runtimeOnly` on them in the flow worker application `build.gradle` (`applications:workers:release:flow-worker`).

omfR3 committed:
Also add net.corda.ledger.models to sandbox whitelist

omfR3 committed:
Try making flow worker depend on corda-ledger project.
Many temporary changes to libs/ledger/models/build.gradle to try to get sandboxed integration test working (it's not).
Add integration test for ConsensualStatesLedger service (based on SecurityManagerTests). Doesn't build yet.

omfR3 commented:
FYI, rebased on top of `release/os/5.0` because we were pretty far behind... the result is a bit ugly so it might have been better to merge and maybe squash later.

JamesHR3 committed:
CORE-5102: Flow marker interface and generic marshalling service (#1447)

* CORE-5102: Add a generic marshalling service and move the JSON marshalling service to use it

* CORE-5102: Introduce common Flow interface and move old Flow to Subflow

* CORE-5102: Add further test and update the CorDapp manifest

* CORE-5102: Add more tests for the new flow implementation classes

* CORE-5102: Move the parseList API down to the MarshallingService

* CORE-5102: Add getRequestBodyAsList method to RPCRequestData

JasonSinclair95 committed:
Core 4600 kafka auth attempt2 (#1439)

* changes made for kafka authentication

omfR3 committed:
Move VirtualNodeService.kt from `:components:virtual-node:sandbox-group-context-service integration tests` into a new subproject `:testing:virtual-node` so it can be re-used elsewhere. Doesn't work yet and breaks the integration tests.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5090/injectable-consensual-ledger

omfR3 committed:
Remove remnant of integration test fixtures (didn't work with OSGi)

omfR3 committed:
(sandbox-group-context-service) Use the new package for VirtualNodeService in integration tests

omfR3 committed:
Try to follow changes in ledger module structure

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5090/injectable-consensual-ledger

omfR3 committed:
More reshuffling to match new ledger-consensual-impl structure

omfR3 committed:
Consensual ledger service probably needs 'immediate = true' in component annotation

omfR3 committed:
Update whitelist in SandboxGroupContextComponentImpl to use new consensual ledger bundle name

omfR3 committed:
Re-enable (broken) injection of ConsensualLedgerService so integration test fails again

omfR3 committed:
Fix consensual ledger injection (by removing a stray package-info.java file).

omfR3 committed:
Cleanup:

* Remove some probably-currently-unneeded requirements from consensual ledger gradle build file + test.bndrun.
* Remove -impl from SandboxGroupContextComponentImpl and test SandboxSetupImpl.

omfR3 committed:
Fix incorrect dependency in flow-worker application (should depend on impl project, rather than interface)

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5090/injectable-consensual-ledger

omfR3 committed:
Slight cleanup

omfR3 committed:
Expand testing/virtual-node/.../VirtualNodeService.kt to try loading a SubFlow first, then an RPCStartableFlow. Injection within the flow fails, though.

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5090/injectable-consensual-ledger

omfR3 committed:
Revert "Expand testing/virtual-node/.../VirtualNodeService.kt to try loading a SubFlow first, then an RPCStartableFlow. Injection within the flow fails, though."

This reverts commit cfd766eeef36989b30a3edfe8eea7137d419cf07.

omfR3 committed:
ConsensualLedgerServiceTest: Don't use VirtualNodeService to start a flow because injection won't work yet.

omfR3 committed:
Temporary build version to match changes in corda-api

barrylapthorn commented on testing/cpbs/consensual-ledger/build.gradle:
```diff
@@ -0,0 +1,19 @@
+plugins {
```
Any chance this can be merged into the `extensible-cpb` project instead?

Overall "we" would like to reduce the number of cpbs we're using for testing and the number of gradle projects to keep the compile times down.

If not, no worries.

omfR3 replied:
Hey @barrylapthorn, that sounds like a good idea... would there be any negative consequences to doing it?

barrylapthorn replied:
I don't think so.  Entirely up to you if you want to do that.  It probably helps "us" in the sense that a larger cpb with more flows might well flush out problems or test code paths we don't normal visit.

omfR3 replied:
Makes sense. @blsemo should we do this now or would it interfere with your work on the system flows? If you've not touched the CPB, I'm happy to try moving it into the `extensible-cpb` project (although I just **know** OSGi will come and slap me around somehow -- it always does).

blsemo replied:
Moving it is fine.

omfR3 replied:
Hi @barrylapthorn -- I had a look at the current `extendable-cpb` project and am a bit confused how to take advantage of it. I see a few integration tests extending the CPB, and `extendable-cpb/build.gradle` itself references two other projects:
```
    cordapp project(':testing:bundles:testing-cats')
    cordapp project(':testing:bundles:testing-dogs')
```
There's no `src` directory in the project, but at the bottom of `build.gradle` is this:
```
    // TODO - add flow CPK here
```
Should I just move the `src` dir for the consensual flow test CPB into `extendable-cpb` and add the required dependencies into `build.gradle`? I'm not sure if that's what you intended, but I guess it would allow us to get rid of `testing/cpbs/consensual-ledger`. But then if everybody does that, it'll become a gravity well full of every possible dependency...

barrylapthorn replied:
Maybe it's just better to leave it as-is then.

Rather than including the _cpk_ code "inline" in the `src` folder within the *cpb* (i.e. this `build.gradle` with the `cpb` plugin), you could move the source to it's own "cpk" project, then include it in the cpb, i.e.

```
plugins {
    id 'net.corda.plugins.cordapp-cpb'
    id 'org.jetbrains.kotlin.jvm'
}

description 'Consensual flow CPB'
group 'com.r3.testing.consensual-flow'

cordapp {
    targetPlatformVersion platformVersion as Integer
    minimumPlatformVersion platformVersion as Integer
    workflow {
        name "Consensual flow"
        versionId 1
        vendor "R3"
    }
}

dependencies {
    cordaProvided platform("net.corda:corda-api:$cordaApiVersion")
    cordaProvided 'net.corda:corda-application'

    cordapp project(':testing:bundles:consensual-flow')

    //     cordapp project(':testing:bundles:other-flow')
    //     cordapp project(':testing:bundles:some-contract')
}

```

This then makes flows reusable between cpbs (just like regular jars).

But like I said above, maybe just leave it as-is for now.

omfR3 commented:
Closing as superceded by other PRs.
